src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -245,56 +261,64 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
+        private Activity? StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
         {
-            var activity = new Activity(ActivityName);
             hasDiagnosticListener = false;
+            if (_diagnosticListener.IsEnabled(ActivityStartKey))
+            {
+                hasDiagnosticListener = true;
+            }
+
+            // Short-circuit to avoid doing an expensive header lookup
+            if (!(hasDiagnosticListener || _activitySource.HasListeners()))
+            {
+                return null;
+            }
 
             var headers = httpContext.Request.Headers;
             if (!headers.TryGetValue(HeaderNames.TraceParent, out var requestId))
             {
                 headers.TryGetValue(HeaderNames.RequestId, out requestId);
             }
-
-            if (!StringValues.IsNullOrEmpty(requestId))
+            var activity = _activitySource.StartActivity(ActivityName, ActivityKind.Server, requestId);
```

Comment:
>var [](start = 12, length = 3)

We have this request we are planning to support in 6.0 [https://github.com/dotnet/runtime/issues/42784](https://github.com/dotnet/runtime/issues/42784) which would make this pattern much easier I guess.